### PR TITLE
Ability to filter posts containing keyword in title/body

### PR DESCRIPTION
**Pull request by hjiangsu**
## Pull Request Description

> Review with whitespace off

This PR enables the ability to add and remove keywords. When adding keywords, any post that contains the keyword in the title or body will be filtered out from the feed page (and any widget that uses the feed page).

Some additional notes:
- Added `width: min(MediaQuery.of(context).size.width, 700)` to `showThunderDialog`'s `AlertDialog`. This allows alert dialogs to take up more of the width of the screen similar to the input dialogs (community, user, instance, etc)
- Applied some small UI tweaks to `UserSettingsPage` since it now appears on the Settings page, I've adjusted some of the UI for blocking users/communities/instances to match the other UIs present in the Settings page.

## Issue Being Fixed

Issue Number: #513

## Screenshots / Recordings

https://github.com/thunder-app/thunder/assets/30667958/639c179d-8b96-4a50-95e2-847ee87bf446

## Checklist

- [x] Did you update CHANGELOG.md?
- [x] Did you use localized strings where applicable?
- [x] Did you add `semanticLabel`s where applicable for accessibility?

### CODE REVIEW

**Comment by hjiangsu** [comment]
@micahmo ready for review!

File: lib/user/pages/user_settings_page.dart [hunk 1]
```diff
@@ -163,25 +163,28 @@ class _UserSettingsPageState extends State<UserSettingsPage> {
...
-                          onPressed: () => showInstanceInputDialog(
-                            context,
-                            title: l10n.blockInstance,
-                            onInstanceSelected: (instance) {
-                              context.read<UserSettingsBloc>().add(UnblockInstanceEvent(instanceId: instance.id, unblock: false));
-                            },
-                          ),
+                      Padding(
+                        padding: const EdgeInsets.symmetric(horizontal: 16.0, vertical: 8.0),
+                        child: Row(
```
Comment (micahmo) [inline lib/user/pages/user_settings_page.dart:169]: The only tiny problem with these changes is that the plus button (to add a new block) and the x button (to delete a block) are no longer aligned. It's very minor though. 😊

**Review by micahmo — APPROVED** [review]
Awesome feature!

One thing I noticed while testing... This might be a problem I introduced with my dialog refactor, but I was able to add an empty string as a filter by pressing enter on the keyboard. (I think this is a problem with the comment search too.)

**Comment by hjiangsu** [comment]
> I  was able to add an empty string as a filter by pressing enter on the keyboard

Ahh, I missed that edge case while testing it. I can add in an additional check to make sure we can't add in empty strings (and also trim whitespaces while I'm at it)

File: lib/feed/utils/post.dart [hunk 2]
```diff
@@ -39,6 +46,16 @@ Future<Map<String, dynamic>> fetchPosts({
     // Remove deleted posts
     getPostsResponse = getPostsResponse.copyWith(posts: getPostsResponse.posts.where((PostView postView) => postView.post.deleted == false).toList());
 
+    // Remove posts that contain any of the keywords in the title or body
+    getPostsResponse = getPostsResponse.copyWith(
+      posts: getPostsResponse.posts.where((postView) {
+        final title = postView.post.name.toLowerCase();
+        final body = postView.post.body?.toLowerCase() ?? '';
+
+        return !keywordFilters.any((keyword) => title.contains(keyword.toLowerCase()) || body.contains(keyword.toLowerCase()));
```
[the remark this answers is not in the document]
Comment (hjiangsu) [reply lib/feed/utils/post.dart:55]: Yeah, I'm hoping this isn't slow either 😅 This is the side-effect of Lemmy not having an internal implementation for keyword filtering so I hope they add it in sometime in the future to reduce the need for this!

File: lib/settings/pages/filter_settings_page.dart [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            ),
+          ),
+          SliverToBoxAdapter(
+            child: Padding(
+              padding: const EdgeInsets.only(left: 16.0, right: 16.0),
+              child: Align(
+                alignment: Alignment.centerLeft,
+                child: keywordFilters.isEmpty
+                    ? Text(
+                        l10n.noKeywordFilters,
```
[the remark this answers is not in the document]
Comment (hjiangsu) [reply lib/settings/pages/filter_settings_page.dart:110]: I was trying to match it with the other placeholder text we have, but I understand what you mean here 😅

[the remark this answers is not in the document]
Comment (hjiangsu) [reply lib/settings/pages/filter_settings_page.dart:110]: I moved the description for the filter to the top of the page. Hopefully this does the trick! b3aa2eb11f39559797c9815899af33c1915aee47

File: lib/settings/pages/filter_settings_page.dart [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                                context: context,
+                                title: l10n.removeKeywordFilter,
+                                contentText: l10n.removeKeyword(keywordFilters[index]),
+                                primaryButtonText: l10n.remove,
+                                onPrimaryButtonPressed: (context, setPrimaryButtonEnabled) {
+                                  setPreferences(LocalSettings.keywordFilters, keywordFilters.where((element) => element != keywordFilters[index]).toList());
+                                  Navigator.of(context).pop();
+                                },
+                                secondaryButtonText: l10n.cancel,
+                                onSecondaryButtonPressed: (context) => Navigator.of(context).pop(),
```
[the remark this answers is not in the document]
Comment (hjiangsu) [reply lib/settings/pages/filter_settings_page.dart]: Done! 8dc3efd9da72921a34281bbc102a4c79a0c5f440

**Comment by hjiangsu** [comment]
Added trimming to `showInputDialog` to disable the button when there's empty strings, and added trimming to keywords in eff689a5f0a4cc4ddd29df89def6389125dedd46!

File: lib/user/pages/user_settings_page.dart (hunk 1 above)
Comment (hjiangsu) [reply lib/user/pages/user_settings_page.dart:169]: I think this should be okay for now! I do think the plus button now lines up with the ToggleOptions so maybe its just a matter of adjusting the x button to be closer to the right

**Comment by micahmo** [comment]
> Added trimming to `showInputDialog` to disable the button when there's empty strings, and added trimming to keywords in [eff689a](https://github.com/thunder-app/thunder/commit/eff689a5f0a4cc4ddd29df89def6389125dedd46)!

Nice! We can try to add something similar to comment search at some point.

File: lib/settings/pages/filter_settings_page.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (micahmo) [reply lib/settings/pages/filter_settings_page.dart:110]: Yeah this looks better!

File: lib/user/pages/user_settings_page.dart (hunk 1 above)
Comment (micahmo) [reply lib/user/pages/user_settings_page.dart:170]: > maybe its just a matter of adjusting the x button to be closer to the right

Agreed!

**Review by micahmo — APPROVED** [review]

**Pull request merged**